heron/common/src/java/com/twitter/heron/common/basics/ByteAmount.java

```diff
@@ -0,0 +1,247 @@
+   *
+   * @param megabytes value in megabytes to represent
+   * @return a ByteAmount object repressing the number of MBs passed
+   */
+  public static ByteAmount fromMegabytes(long megabytes) {
+    if (megabytes >= MAX_MB) {
+      return new ByteAmount(Long.MAX_VALUE);
+    } else {
+      return new ByteAmount(megabytes * MB);
+    }
+  }
+
+  /**
+   * Creates a ByteAmount value in gigabytes. If the gigabytes value
+   * is >= Long.MAX_VALUE / 1024 / 1024 / 1024, the byte representation is capped at Long.MAX_VALUE.
+   *
+   * @param gigabytes value in gigabytes to represent
+   * @return a ByteAmount object repressing the number of GBs passed
+   */
+  public static ByteAmount fromGigabytes(long gigabytes) {
+    if (gigabytes >= MAX_GB) {
+      return new ByteAmount(Long.MAX_VALUE);
+    } else {
+      return new ByteAmount(gigabytes * GB);
+    }
+  }
+
+  /**
+   * Returns the number of bytes associated with the ByteValue.
+   * @return number of bytes
+   */
+  public long asBytes() {
+    return bytes;
+  }
+
+  /**
+   * Converts the number of bytes to megabytes, rounding if there is a remainder. Because of loss
+   * of precision due to rounding, it's strongly advised to only use this method when it is certain
+   * that megabytes were using during the construction of the object via fromMegabytes. Only In that
+   * scenario is there a guarantee that precision will not be lost when calling this method.
+   * @return returns the ByteValue in MBs or 0 if the value is < (1024 * 1024) / 2
+   */
+  public long asMegabytes() {
+    return Math.round((double) bytes / MB);
+  }
+
+  /**
+   * Converts the number of bytes to gigabytes, rounding if there is a remainder. Because of loss
+   * of precision due to rounding, it's strongly advised to only use this method when it is certain
+   * that gigabytes were using during the construction of the object via fromGigabytes. Only In that
+   * scenario is there a guarantee that precision will not be lost when calling this method.
+   * @return returns the ByteValue in GBs or 0 if the value is < (1024 * 1024 * 1024) / 2
+   */
+  public long asGigabytes() {
+    return Math.round((double) bytes / GB);
+  }
+
+  /**
+   * Convenience methdod to determine if byte value is zero
+   * @return true if the byte value is 0
+   */
+  public boolean isZero() {
+    return ZERO.equals(this);
+  }
+
+  /**
+   * Subtracts other from this.
+   * @param other ByteValue to subtract
+   * @return this ByteValue minus other ByteValue
+   * @throws IllegalArgumentException if subtraction would overshoot Long.MIN_VALUE
+   */
+  public ByteAmount minus(ByteAmount other) {
+    if (Long.MIN_VALUE + other.asBytes() > asBytes()) {
+      throw new IllegalArgumentException(String.format(
+          "Subtracting %s from %s would overshoot Long.MIN_LONG", other, this));
+    }
+    return ByteAmount.fromBytes(asBytes() - other.asBytes());
```

Comment:
The method is returning a new object. However, this object should be reduced as per the javadoc?